### PR TITLE
Deep copy object passed to update

### LINKED ISSUE #1059: [BUG] Model.create mutates object when there are nested object properties not in schema
### Summary:

`Model.create` mutates the object when there are nested object properties not in schema.

#### General
```
it("Should not mutate original object if properties not in schema", async () => {
	createItemFunction = () => Promise.resolve();
	const obj = {"id": 1, "random": "data"};
	await callType.func(User).bind(User)(obj);
	expect(obj).to.be.an("object");
	expect(obj).to.eql({
		"id": 1,
		"random": "data"
	});
});

it("Should not mutate original object if nested object properties not in schema", async () => {
	createItemFunction = () => Promise.resolve();
	const obj = {"id": 1, "randomdata": {"hello": "world"}};
	await callType.func(User).bind(User)(obj);
	expect(obj).to.be.an("object");
	expect(obj).to.eql({
		"id": 1,
		"randomdata": {
			"hello": "world"
		}
	});
});
```

### Current output and behavior (including stack trace):

The tests above to pass.

### Expected output and behavior:

The first test passes, the second does not.

### Environment:

Operating System: macOS
Operating System Version: 11.1
Node.js version (`node -v`): v14.0.0
NPM version: (`npm -v`): 6.14.4
Dynamoose version: 2.4.1

### Other information (if applicable):

Steps tried:

1. Created a `utils.object.copy` method which is bascially `return JSON.parse(JSON.stringify(object)); `.
2. Inside `Document.objectFromSchema` I tried using that instead of `const returnObject = {...object};`. That seemed to fix the test, but created a bunch of problems related to symbols and `dynamoose.UNDEFINED`.

### Other:
- [x] I have read through the Dynamoose documentation before posting this issue
- [x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x] I have tested the code provided and am confident it doesn't work as intended
- [x] I have filled out all fields above
- [x] I am running the latest version of Dynamoose

### LINKED ISSUE #1163: [BUG] Timestamps are Date while local object is mutated
### Summary:
Timestamps are in epochs (numbers), but when calling `save()`, the mutated object gets a `Date` object instead. The resulting object returned in the promise has number timestamps again.

#### Schema
```
N/A
```

#### Model
```
const m = dynamoose.model("Foo", new dynamoose.Model{id: String}, {timestamps: true});
const foo1 = await m.create({id: 'abc'});
// now,foo1.updatedAt is a number
console.log('foo1 after create:', foo1);

foo1.hello = 'world'; 
const foo2 = await foo1.save();
// now, foo1.updatedAt is a Date object, but foo2.updatedAt is a number
console.log('foo1 after mutation:', foo1);
console.log('foo2 after mutation:', foo2);
```

#### General
```
N/A
```

### Current output and behavior (including stack trace):

```
foo1 after create: {
  "id": "abc",
  "createdAt": 1616084000594,
  "updatedAt": "1616084000594",
}
foo1 after mutation: {
  "id": "abc",
  "createdAt": 1616084000594,
  "updatedAt": "2021-03-18T16:13:20.646Z",
}
foo2 after mutation: {
  "id": "abc",
  "createdAt": 1616084000594,
  "updatedAt": 1616084000646,
}
```

### Expected output and behavior:

Timestamps should be consistently typed even for mutated local objects that aren't saved yet.

### Environment:

Operating System: Ubuntu
Operating System Version: 18.04
Node.js version (`node -v`): 14.2.0
NPM version: (`npm -v`): 6.14.4
Dynamoose version: 2.7.3

### Other:
- [X] I have read through the Dynamoose documentation before posting this issue
- [X] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [X] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [X] I have tested the code provided and am confident it doesn't work as intended
- [X] I have filled out all fields above
- [X] I am running the latest version of Dynamoose

### LINKED ISSUE #1257: [BUG] update deletes the key from passed object
### Summary:

`update(doc)` deletes the primary key from doc.

#### General
```
import * as dynamoose from "dynamoose";

const Model = dynamoose.model("dynamoose-test", {"id": String, 'name': String});
Model.create({'id': '12', 'name': 'me'}).then((res) => {
    Model.get('12').then((doc1) => {
        doc1.name = 'you'
        console.log(doc1)
        Model.update(doc1).then((doc2) => {
            console.log(doc1)
        })
    })
})
```

### Current output and behavior (including stack trace):

> Document { id: '12', name: 'you' }
> Document { name: 'you' }

### Expected output and behavior:

> Document { id: '12', name: 'you' }
> Document { id: '12', name: 'you' }

### Environment:

Operating System: Ubuntu
Operating System Version: 20.04
Node.js version (`node -v`): tried v14.17.4 and v16.2.0
NPM version: (`npm -v`): 7.20.5
Dynamoose version: 2.8.1

### Other:
- [ x] I have read through the Dynamoose documentation before posting this issue
- [ x] I have searched through the GitHub issues (including closed issues) and pull requests to ensure this issue has not already been raised before
- [ x] I have searched the internet and Stack Overflow to ensure this issue hasn't been raised or answered before
- [x ] I have tested the code provided and am confident it doesn't work as intended
- [ x] I have filled out all fields above
- [ x] I am running the latest version of Dynamoose

### PR DESCRIPTION
**Pull request by andrewda**
### Summary:

See summaries below per-issue.

#### Closes #1257

Previously, updating a document would cause the hash and range keys on the passed object to be deleted. This PR uses `deep_copy` to copy the passed object and ensure the original does not get mutated.

```ts
it("Should not delete keys from object", async () => {
	const obj = {"id": 1, "address": {"zip": 12345, "country": "world"}};

	User = dynamoose.model("User", new dynamoose.Schema({"id": Number, "address": Object}, {"saveUnknown": true}));
	updateItemFunction = () => Promise.resolve({"Attributes": {"id": {"N": "1"}, "address": {"M": {"zip": {"N": "12345"}, "country": {"S": "world"}}}}});
	await callType.func(User).bind(User)(obj);
	expect(obj).to.deep.eql({"id": 1, "address": {"zip": 12345, "country": "world"}});
});
```

#### Closes #1163 

Previously, re-saving a Document would cause the `updatedAt` property to be a `Date` (rather than a numeric epoch time like the others). This PR ensures the type of `updatedAt` and `createdAt` properties are consistently numeric, rather than Date objects.

```ts
it("Should return correct result after saving with timestamps", async () => {
	createItemFunction = () => Promise.resolve();
	User = dynamoose.model("User", new dynamoose.Schema({"id": Number, "name": String}, {"timestamps": true}));
	const date1 = Date.now();
	const obj = {"id": 1, "name": "Charlie"};
	const result = await callType.func(User).bind(User)(obj);

	// Check original timestamps
	expect(result.id).to.eql(1);
	expect(result.name).to.eql("Charlie");
	expect(result.createdAt).to.be.a("number");
	expect(result.createdAt).to.be.within(date1 - 10, date1 + 10);
	expect(result.updatedAt).to.be.a("number");
	expect(result.updatedAt).to.be.within(date1 - 10, date1 + 10);

	await new Promise((resolve) => setTimeout(resolve, 20));

	const date2 = Date.now();

	// Mutate document and re-save
	result.name = "Charlie 2";
	const result2 = await result.save();

	expect(result.toJSON()).to.eql(result2.toJSON());
	[result, result2].forEach((r) => {
		expect(r.id).to.eql(1);
		expect(r.name).to.eql("Charlie 2");
		expect(r.createdAt).to.be.a("number");
		expect(r.createdAt).to.be.within(date1 - 10, date1 + 10);
		expect(r.updatedAt).to.be.a("number");
		expect(r.updatedAt).to.be.within(date2 - 10, date2 + 10);
	});
});
```

#### Closes #1059

Previously, creating a document from an object with deep/nested non-schema properties would mutate the passed object. This PR updates this behavior to leave the passed object untouched, and only mutate the returned document.

```ts
it("Should not mutate original object if properties not in schema", async () => {
	createItemFunction = () => Promise.resolve();
	const obj = {"id": 1, "random": "data"};
	await callType.func(User).bind(User)(obj);
	expect(obj).to.be.an("object");
	expect(obj).to.deep.eql({
		"id": 1,
		"random": "data"
	});
});

it("Should not mutate original object if nested object properties not in schema", async () => {
	createItemFunction = () => Promise.resolve();
	const obj = {"id": 1, "randomdata": {"hello": "world"}};
	await callType.func(User).bind(User)(obj);
	expect(obj).to.be.an("object");
	expect(obj).to.deep.eql({
		"id": 1,
		"randomdata": {
			"hello": "world"
		}
	});
});
```

### Type (select 1):
- [x] Bug fix
- [ ] Feature implementation
- [ ] Documentation improvement
- [ ] Testing improvement
- [ ] Something not listed here

### Is this a breaking change? (select 1):
- [ ] 🚨 YES 🚨
- [x] No
- [ ] I'm not sure

### Is this ready to be merged into Dynamoose? (select 1):
- [x] Yes
- [ ] No

### Are all the tests currently passing on this PR? (select 1):
- [x] Yes
- [ ] No

### Other:
- [x] I have read through and followed the Contributing Guidelines
- [x] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [x] I have updated the Dynamoose documentation (if required) given the changes I made
- [x] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [x] I have ensured the following commands are successful from the root of the project directory
  - [x] `npm test`
  - [x] `npm run lint`
- [x] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoose/dynamoose/blob/master/LICENSE)
- [x] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [x] I have filled out all fields above

### CODE REVIEW

File: test/unit/Model.js [hunk 1]
```diff
@@ -4717,6 +4717,12 @@ describe("Model", () => {
 
 				expect(result).to.eql("Dynamoose Warning: Passing callback function into transaction method not allowed. Removing callback function from list of arguments.");
 			});
+
+			it("Should not delete keys from object", () => {
```
Comment (fishcharlie) [inline test/unit/Model.js:4721]: Looks like this is a `transaction.update` call. I think we should add one for just `update` as well.

File: lib/utils/deep_copy.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	if (obj instanceof Array) {
+		copy = obj.map((i) => deep_copy(i));
+		return copy;
+	}
+
+	// Handle Object
+	if (obj instanceof Object) {
+		copy = {};
+		for (const attr in obj) {
+			if (Object.prototype.hasOwnProperty.call(obj, attr)) copy[attr] = deep_copy(obj[attr]);
```
Comment (fishcharlie) [inline lib/utils/deep_copy.ts:21]:
```suggestion
			if (Object.prototype.hasOwnProperty.call(obj, attr)) {
				copy[attr] = deep_copy(obj[attr]);
			}
```

**Review by fishcharlie — CHANGES_REQUESTED** [review]
This looks pretty good. I wonder if this issue is also related to #1163 or #1059. Not sure if it is or not.

File: test/unit/Model.js (hunk 1 above)
Comment (andrewda) [reply test/unit/Model.js:4721]: Ah, good point!

**Comment by andrewda** [comment]
Added the additional `Mode.update` test! Also tested on master and, as expected, all tests are failing:

```
1) Model
       Model.update
         Promise
           Should not delete keys from object:

      AssertionError: expected { Object (address) } to deeply equal { Object (id, address) }
      + expected - actual

         "address": {
           "country": "world"
           "zip": 12345
         }
      +  "id": 1
       }
      
      at Context.<anonymous> (test/unit/Model.js:3560:26)

  2) Model
       Model.update
         Callback
           Should not delete keys from object:

      AssertionError: expected { Object (address) } to deeply equal { Object (id, address) }
      + expected - actual

         "address": {
           "country": "world"
           "zip": 12345
         }
      +  "id": 1
       }
      
      at Context.<anonymous> (test/unit/Model.js:3560:26)

  3) Model
       Model.transaction
         Model.transaction.update
           Should not delete keys from object:

      AssertionError: expected { name: 'Bob' } to deeply equal { id: 1, name: 'Bob' }
      + expected - actual

       {
      +  "id": 1
         "name": "Bob"
       }
```

**Comment by andrewda** [comment]
> I wonder if this issue is also related to #1163 or #1059. Not sure if it is or not.

Yep, could be! Although this PR won't address those since this only touches `.update`. I'll take a look in a few hours to see if I can get this to address those other issues.

**Comment by andrewda** [comment]
@fishcharlie I've updated this PR to fix both #1163 and #1059. Check the description for updated details!

**Comment by fishcharlie** [comment]
@andrewda Looks good overall. Is there **any** way that we can remove the dependency on lodash deep clone? Really have the goal of trying to reduce the amount of dependencies in this project as much as possible. I'm sure there was a reason for using it in the first place, instead of your custom function. So just wanna see what we can do about that.

**Comment by andrewda** [comment]
@fishcharlie I've tried to write my own deep clone a few times now, but the following test always seems to fail. Tried digging through to see what could be causing this for an hour and came up empty, though I suspect it has something to do with cloning Documents.

```
  1) Document
       document.populate
         Promise
           Should not call getItem if sub document already exists:

      AssertionError: expected [ Array(1) ] to deeply equal []
      + expected - actual

      -[
      -  {
      -    "Key": {
      -      "id": {
      -        "N": "1"
      -      }
      -    }
      -    "TableName": "User"
      -  }
      -]
      +[]
```

Lodash's deep clone function handles this and other edge cases well, and I'd trust it way more than something I were to write myself. If adding the additional dependency is really out of the question, I can continue trying to write our own native one, but not sure how long it will take for me to get something working.

**Comment by fishcharlie** [comment]
@andrewda What about something like 6d2ccfe5bfc1aa96d0299d22065662959f9c74f0?

**Comment by fishcharlie** [comment]
Tests are failing due to [`Object.fromEntries`](https://node.green/#ES2019-features--Object-fromEntries) only being added in Node.js v12. If @andrewda thinks 6d2ccfe is a good change, I can try to find some time to find a workaround for Node.js versions less than v12.

**Comment by andrewda** [comment]
> Tests are failing due to [`Object.fromEntries`](https://node.green/#ES2019-features--Object-fromEntries) only being added in Node.js v12. If @andrewda thinks [6d2ccfe](https://github.com/dynamoose/dynamoose/commit/6d2ccfe5bfc1aa96d0299d22065662959f9c74f0) is a good change, I can try to find some time to find a workaround for Node.js versions less than v12.

Looks perfect!

**Comment by andrewda** [comment]
@fishcharlie I just pushed c8eab5d38f654745a7f5d0cd1320914cb34a56e4 which should resolve the `Object.fromEntries` issue, and add a new test for [one line](https://github.com/dynamoose/dynamoose/pull/1258/files#diff-d193292759d2635b89f2e7054f865c11157828de4008c218258a4bdff4edd2afR44) which wasn't tested.

**Review by fishcharlie — APPROVED** [review]

**Comment by fishcharlie** [comment]
Ok @andrewda. I made one more change to this. 483706df53e248eeed1f87ad468343bcc488d645. It should improve how we are copying object instances.

Let me know what you think.

**Comment by andrewda** [comment]
@fishcharlie Ah yeah that looks perfect. LGTM :+1:

**Pull request merged**